helix-term/src/ui/popup.rs

```diff
@@ -163,7 +165,40 @@ impl<T: Component> Popup<T> {
             },
         };
 
-        match final_pos {
+        // compute maximum space available for child
+        let mut max_height = match final_pos {
+            Open::Above => rel_y,
+            Open::Below => viewport.height.saturating_sub(1 + rel_y),
+        };
+        max_height = max_height.min(MAX_HEIGHT);
+        let mut max_width = viewport.width.saturating_sub(2).min(MAX_WIDTH);
+        render_borders = render_borders && max_height > 3 && max_width > 3;
+        if render_borders {
+            max_width -= 2;
+            max_height -= 2;
+        }
+
+        // compute required child size and reclamp
+        let (mut width, child_height) = self
+            .contents
+            .required_size((max_width, max_height))
+            .expect("Component needs required_size implemented in order to be embedded in a popup");
+
```

Comment:
```suggestion

        if max_height < child_height && final_pos == Open::Below && rel_y > max_height {
            final_pos = Open::Above;
        }
        

```
For documentation popups, i would like to see as much information as possible, so if it doesnt fit below, and we have more space above, then display it above.
Does not affect menus.

Reply:
That was part of the previous PR that changed the popup positioning. I really don't want to start adding edgecases like that. The whole special case for menu thing is extremely temporary and something I want to get rid of and instead of lean on more